sky/provision/__init__.py

```diff
@@ -141,10 +141,15 @@ def query_ports(
     provider_name: str,
     cluster_name_on_cloud: str,
     ports: List[str],
+    head_ip: Optional[str] = None,
     provider_config: Optional[Dict[str, Any]] = None,
 ) -> Dict[int, List[common.Endpoint]]:
     """Query details about ports on a cluster.
 
+    If head_ip is provided, it may be used by the cloud implementation to
+    return the endpoint without querying the cloud provider. If head_ip is not
+    provider, the cloud provider will be queried to get the endpoint info.
+
     Returns a dict with port as the key and a list of common.Endpoint.
     """
     raise NotImplementedError
```

Comment:
How about use the implementation here to let the API to use the default implementation in this file if it is not implemented in the specific provider?

https://github.com/skypilot-org/skypilot/blob/421e62402ce51576a5668ce139234a26c52f7475/sky/provision/__init__.py#L29-L50

Reply:
Hmm, not sure if I fully follow - do you have an example in mind? Do you mean:
1. Calling `common.query_ports_passthrough(ports, head_ip)` in query_ports here
2. Modify `_route_to_cloud_impl` to route only if underlying provider implements the method?

Reply:
Good idea - switched the implementation.